Add Moehlenhoff Alpha2 underfloor heating system integration

j-a-n opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add Moehlenhoff Alpha2 underfloor heating system support.
[Link to manufacturer](https://www.moehlenhoff.de/oem-alpha-2-system)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/21328

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

j-a-n committed:
Add Moehlenhoff Alpha2 underfloor heating system integration

homeassistant commented:
Hi @j-a-n, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#42771).

Once you do that we will be able to review and accept this pull request.

Thanks!

j-a-n committed:
isort changes

j-a-n committed:
flake8 changes

j-a-n committed:
Do not exclude config_flow.py

j-a-n committed:
pylint changes

j-a-n committed:
Add config_flow test

j-a-n committed:
correct requirements_test_all.txt

j-a-n committed:
more tests

j-a-n committed:
Merge branch 'dev' into moehlenhoff-alpha2

j-a-n committed:
Update test description

j-a-n committed:
Test connection and catch TimeoutError in async_setup_entry

j-a-n committed:
Add version to manifest file

j-a-n committed:
Remove version from manifest file

j-a-n committed:
Replace tests.async_mock.patch by unittest.mock.patch

j-a-n committed:
Update moehlenhoff-alpha2 to version 1.0.1

j-a-n committed:
Update requirements for moehlenhoff-alpha2 1.0.1

j-a-n committed:
Update moehlenhoff-alpha2 to 1.0.2

eavanvalkenburg commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,124 @@
+    base_uh = Alpha2BaseUpdateHandler(base, 60)
+    hass.data.setdefault(DOMAIN, {"connections": {}, "devices": set()})
+    hass.data[DOMAIN]["connections"][entry.entry_id] = base_uh
+    for component in PLATFORMS:
```
you can use this instead: ```hass.config_entries.async_setup_platforms(entry, PLATFORMS)```

j-a-n replied:
First of all thanks a lot for your review!
Now using async_setup_platforms instead.

eavanvalkenburg commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,124 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
```
and this: ```unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)```

j-a-n replied:
Done.

eavanvalkenburg commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,124 @@
+    except TimeoutError as err:
+        raise exceptions.ConfigEntryNotReady from err
+    base_uh = Alpha2BaseUpdateHandler(base, 60)
+    hass.data.setdefault(DOMAIN, {"connections": {}, "devices": set()})
```
seems these are not linked to this entry_id, why not add this under the entry_id, so each entry keeps track of their own connections and devices

j-a-n replied:
Your right, that makes sense. Done.

eavanvalkenburg commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,124 @@
+        )
+    )
+    if unload_ok and entry.entry_id in hass.data[DOMAIN]:
+        hass.data[DOMAIN].pop(entry.entry_id)
```
this is where the above goes wrong, the entry_id is not present under domain, but under [domain][connections], so fix this so that everything is under the entry_id, and then unpack whatever is in there.

j-a-n replied:
Done.

eavanvalkenburg commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,124 @@
+    async def async_update(self):
+        """Pull the latest data from the Alpha2 base."""
+        # Only update every update_interval
+        if (time.monotonic() - self._last_update) >= self.scan_interval:
```
it's probably easier for a case like this to schedule the next update after this update completed, using ```from homeassistant.helpers.event import async_call_later```, then you also don't have to worry about storing last_update

j-a-n replied:
Using homeassistant.helpers.event.async_track_time_interval now.

eavanvalkenburg commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,179 @@
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return True
```
this should be false when this entity does not do updates itself, like this does!

j-a-n replied:
Returning `False` now and calling async_schedule_update_ha_state after data update.

eavanvalkenburg commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,179 @@
+        """Initialize Alpha2 ClimateEntity."""
+        self._base_update_handler = base_update_handler
+        self._data = data
+        self._base_update_handler.add_heatarea_update_callback(
```
There is a built-in callback registring approach in HA called dispatcher, not sure which component has a clean implementation of this but check it out!

j-a-n replied:
Now using HA dispatcher.

eavanvalkenburg commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,179 @@
+
+    def update(self):
+        """Get latest data from Alpha2 base."""
+        self._base_update_handler.update()
```
do you want each entity to be able to call for an update, if the goal of the central updater is to make sure the api doesn't get overloaded then you don't!

j-a-n replied:
Remove update method.

eavanvalkenburg commented on homeassistant/components/moehlenhoff_alpha2/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+    """Möhlenhoff Alpha2 config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
```
this has moved to manifest, remove it here.

j-a-n replied:
Removed.

eavanvalkenburg commented on homeassistant/components/moehlenhoff_alpha2/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+DATA_SCHEMA = vol.Schema({"host": str})
+
+
+async def validate_input(hass: core.HomeAssistant, data):
```
you're not using hass.

j-a-n replied:
Removed hass.

eavanvalkenburg commented on homeassistant/components/moehlenhoff_alpha2/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+        errors = {}
+        if user_input is not None:
+            try:
+                for entry in self.hass.config_entries.async_entries(DOMAIN):
```
you can use: ```for entry in self._async_current_entries(include_ignore=False):```

j-a-n replied:
Thanks, now using _async_current_entries.

eavanvalkenburg commented on homeassistant/components/moehlenhoff_alpha2/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+                    if entry.data["host"] == user_input["host"]:
+                        return self.async_abort(reason="already_configured")
+
+                info = await validate_input(self.hass, user_input)
```
maybe do this before checking if the host is unique, if the host is wrong then no point in checking uniqueness...

j-a-n replied:
validate_input() sends a http request to the device, so I think the order makes sense.

eavanvalkenburg commented on homeassistant/components/moehlenhoff_alpha2/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+        """Handle a flow initialized by the user."""
+        errors = {}
+        if user_input is not None:
+            try:
```
please wrap this around the least amount of code as possible, maybe even move the try check into the validate function and then return the errors and/or title.

j-a-n replied:
Moved the handling of ClientConnectorError and TimeoutError to validate_input().

milanmeu replied:
There is still too much code in the try block.
Move `except Exception` to validate_input and remove the try here.

milanmeu replied:
Catch `Exception` in `validate_input` so this try can be removed.

j-a-n replied:
Done.

eavanvalkenburg commented on homeassistant/components/moehlenhoff_alpha2/translations/de.json:
```diff
@@ -0,0 +1,19 @@
+{
```
this file is created once it goes in and it gets translated, can be deleted from the PR

j-a-n replied:
Deleted.

eavanvalkenburg commented on homeassistant/components/moehlenhoff_alpha2/translations/en.json:
```diff
@@ -0,0 +1,19 @@
+{
```
same as DE file

j-a-n replied:
Deleted.

eavanvalkenburg commented on tests/components/moehlenhoff_alpha2/test_config_flow.py:
```diff
@@ -0,0 +1,79 @@
+    assert result["type"] == "form"
+    assert result["step_id"] == "user"
+
+    result = await hass.config_entries.flow.async_configure(
```
should really mock the call to create the base with the error, now this will wait during the test, no need for that, mock with side-effect, like below!

j-a-n replied:
Done.

eavanvalkenburg reviewed (changes requested):
overall not too bad, most important change is using dispatcher instead of a self built callback scheme!

j-a-n committed:
Use async_setup_platforms

j-a-n committed:
Use async_unload_platforms

j-a-n committed:
Separate connection and devices for each entry_id

j-a-n committed:
Use async_track_time_interval to schedule updates

j-a-n committed:
Check if input is valid before checking uniqueness

j-a-n committed:
Move Exception handling to validate_input

j-a-n committed:
Catch aiohttp.client_exceptions.ClientConnectorError

j-a-n committed:
Remove translation files

j-a-n committed:
Mock TimeoutError

j-a-n committed:
Merge branch 'home-assistant:dev' into moehlenhoff-alpha2

j-a-n committed:
Merge branch 'moehlenhoff-alpha2' of https://github.com/j-a-n/core into moehlenhoff-alpha2

j-a-n committed:
Fix data update

j-a-n committed:
Merge branch 'dev' into moehlenhoff-alpha2

j-a-n committed:
Replace current callback implementation with ha dispatcher

j-a-n committed:
Return False in should_poll

j-a-n committed:
Remove unused argument

j-a-n committed:
Remove CONNECTION_CLASS

j-a-n committed:
Use _async_current_entries

j-a-n committed:
Call async_schedule_update_ha_state after data update

milanmeu commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,109 @@
+
+class CannotConnect(exceptions.HomeAssistantError):
+    """Error to indicate we cannot connect."""
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up Alpha2 integration."""
+    return True
```
```suggestion
```

milanmeu commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,185 @@
+from . import DOMAIN, SIGNAL_HEATAREA_DATA_UPDATED
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up platform."""
+    return None
```
```suggestion
```

milanmeu commented on homeassistant/components/moehlenhoff_alpha2/config_flow.py:
```diff
@@ -0,0 +1,61 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_SCHEMA = vol.Schema({"host": str})
```
```suggestion
DATA_SCHEMA = vol.Schema({vol.Required("host"): str})
```

milanmeu commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,109 @@
+PLATFORMS = ["climate"]
+
+UPDATE_INTERVAL = timedelta(seconds=60)
+
+
+class CannotConnect(exceptions.HomeAssistantError):
+    """Error to indicate we cannot connect."""
```
Not used?
```suggestion
```

milanmeu commented on homeassistant/components/moehlenhoff_alpha2/manifest.json:
```diff
@@ -0,0 +1,14 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/moehlenhoff_alpha2",
+  "requirements": ["moehlenhoff-alpha2==1.0.2"],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
+  "dependencies": [],
```
```suggestion
  "iot_class": "local_push",
```

j-a-n committed:
Remove unneeded async_setup

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

j-a-n committed:
Remove unneeded async_setup_platform

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

j-a-n committed:
Set Schema attribute host required

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

j-a-n committed:
Remove unused Exception class

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

j-a-n committed:
Update manifest.json

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

j-a-n committed:
pylint constructor return type None

j-a-n commented:
@milanmeu @eavanvalkenburg Could someone please approve the workflows? Thanks!

milanmeu commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,180 @@
+    def set_hvac_mode(self, hvac_mode: str):
+        """Set new target hvac mode."""
+        return
+
+    @property
+    def hvac_action(self):
+        """Return the current running hvac operation if supported."""
+        return None
```
Same as in super class
https://github.com/home-assistant/core/blob/24c1256c2c8bef2d39a2e483ac09689d850181a0/homeassistant/components/climate/__init__.py#L179
https://github.com/home-assistant/core/blob/24c1256c2c8bef2d39a2e483ac09689d850181a0/homeassistant/components/climate/__init__.py#L336-L342
```suggestion
```

j-a-n replied:
Removed.

j-a-n commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,100 @@
+            _LOGGER.debug("Heatarea: %s", heatarea)
+            async_dispatcher_send(self._hass, SIGNAL_HEATAREA_DATA_UPDATED, heatarea)
+
+    def update(self):
+        """Pull the latest data from the Alpha2 base (sync version)."""
+        asyncio.run_coroutine_threadsafe(self.async_update(), self._loop)
```
It's not use anymore, I mill remove it.

j-a-n commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,180 @@
+
+    def set_hvac_mode(self, hvac_mode: str):
+        """Set new target hvac mode."""
+        return
```
Could be replaced with `pass` but not removed.

j-a-n committed:
Replace properties by class variables

j-a-n commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,180 @@
+    """Alpha2 ClimateEntity."""
+
+    target_temperature_step = 0.2
+
```
Okay.

j-a-n committed:
use pass instead of return

j-a-n committed:
Remove unused sync update method

j-a-n commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,180 @@
+        """Initialize Alpha2 ClimateEntity."""
+        self._base_update_handler = base_update_handler
+        self._data = data
+
```
self._data will change over time. Your suggestion would set fixed values for these attributes in the constructor.

milanmeu replied:
Name could indeed change but it seemed unlikely that the min and max temperature would change.
You can keep them as properties.

j-a-n committed:
remove property hvac_action

milanmeu commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,152 @@
+
+    def set_hvac_mode(self, hvac_mode: str):
+        """Set new target hvac mode."""
+        pass
```
We usually leave it empty.

Examples: 

https://github.com/home-assistant/core/blob/24c1256c2c8bef2d39a2e483ac09689d850181a0/homeassistant/components/velbus/climate.py#L83-L84

https://github.com/home-assistant/core/blob/24c1256c2c8bef2d39a2e483ac09689d850181a0/homeassistant/components/xs1/climate.py#L112-L114
https://github.com/home-assistant/core/blob/24c1256c2c8bef2d39a2e483ac09689d850181a0/homeassistant/components/toon/climate.py#L122-L125
```suggestion
```

milanmeu commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,152 @@
+        elif preset_mode == PRESET_ECO:
+            heatarea_mode = 2
+
+        try:
+            await self._base_update_handler.async_set_heatarea_mode(
+                self._data["ID"], heatarea_mode
+            )
+            self._data["HEATAREA_MODE"] = heatarea_mode
+            if heatarea_mode == 1:
+                self._data["T_TARGET"] = self._data["T_HEAT_DAY"]
+            elif heatarea_mode == 2:
+                self._data["T_TARGET"] = self._data["T_HEAT_NIGHT"]
+            return True
+        except Exception as update_err:  # pylint: disable=broad-except
+            _LOGGER.error("Setting target temperature failed: %s", update_err)
+            return False
```
- `async_set_preset_mode` should not return
- there is too much code in the `try` block
- `Exception` is probably too broad
    - It may be that hass can't connect to the device, log a nicer error in this case instead of `update_err`

j-a-n replied:
Okay, comitted some changes.

j-a-n committed:
remove pass

j-a-n committed:
rework exception handling

j-a-n commented:
Hello @milanmeu and thanks for your review! Do you have further suggestions? If you're review is finished, would you mind to start the workflows and update the PR status?
Thank you!

milanmeu commented on homeassistant/components/moehlenhoff_alpha2/config_flow.py:
```diff
@@ -0,0 +1,61 @@
+        """Handle a flow initialized by the user."""
+        errors = {}
+        if user_input is not None:
+            try:
+                for entry in self._async_current_entries(include_ignore=False):
+                    if entry.data["host"] == user_input["host"]:
+                        return self.async_abort(reason="already_configured")
+
```
```suggestion
            self._async_abort_entries_match({"host": user_input["host"]})
                try:
```

j-a-n committed:
Update homeassistant/components/moehlenhoff_alpha2/config_flow.py

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

j-a-n committed:
Correct indentation

j-a-n committed:
catch Exception in validate_input

j-a-n commented:
Hello @milanmeu, would it be possible to start the workflows and update the status?

j-a-n committed:
Merge branch 'home-assistant:dev' into moehlenhoff-alpha2

j-a-n commented:
Link to brands pull request: https://github.com/home-assistant/brands/pull/2757

milanmeu commented on tests/components/moehlenhoff_alpha2/test_config_flow.py:
```diff
@@ -0,0 +1,83 @@
+from moehlenhoff_alpha2 import Alpha2Base
+
+from homeassistant.components.moehlenhoff_alpha2.const import DOMAIN
+from homeassistant.helpers.typing import HomeAssistantType
```
Change `HomeAssistantType` to `HomeAssistant`:
https://github.com/home-assistant/core/blob/c74f9a8313c621c9d8c777b07216c286d55d6118/homeassistant/helpers/typing.py#L28-L33

j-a-n replied:
Done.

j-a-n committed:
Replace HomeAssistantType with HomeAssistant

j-a-n committed:
Update to moehlenhoff-alpha2 1.0.3

j-a-n committed:
Allow to switch between heating and cooling mode

j-a-n committed:
Update moehlenhoff-alpha2 to version 1.0.4

j-a-n committed:
Update heatarea data after setting target temperature

j-a-n committed:
Support hvac_action

j-a-n committed:
Fix heatarea update with multiple bases

j-a-n committed:
Update data after setting preset mode

j-a-n committed:
Use custom preset modes like defined by device

j-a-n commented:
@MartinHjelmare Since I accidentally closed and then reopened this PR some months ago, it stays in cancelled in New Integrations. Could you please move it back to incoming? Thanks.

j-a-n committed:
Merge branch 'dev' into moehlenhoff-alpha2

andyboeh commented:
Hey,

I was made aware of this PR by a community member as I am an author of an addon for the Alpha 2 (ezr2mqtt, going the MQTT route). I have yet to try your component, but one thing I noticed is the lack of an undocumented "feature" of the Alpha 2: You can turn off a heat area using the room controller. If you do that, the reported heatarea state is 3. I was in touch with Möhlenhoff support and they confirmed that this state can only be set and reset via a room controller, not via XML.

One feature that I have integrated on request by a community member is reporting the percentage of the individual actors. I do that by representing every actor by a sensor entity. Maybe we can add something similar to your component?

j-a-n committed:
Fix config flow test

j-a-n committed:
Fix test_duplicate_error

j-a-n committed:
Rename property to extra_state_attributes

Rename property device_state_attributes to extra_state_attributes and
return lowercase keys in dict.

j-a-n commented:
Hi @andyboeh ,
thanks for the hint.
Are you talking about the attribute _HEATAREA_MODE_?
This attribute is set to 3 if I turn off a room controller.

I am trying to get this merged since over one year, so I don't want to add any new features in this PR.
But once this is merged, I would be happy to work with you on new features.

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,108 @@
+            target_temperature,
+        )
+        await self.base.update_heatarea(heatarea_id, {"T_TARGET": target_temperature})
+        await self.async_update()
```
Refresh the coordinator.

j-a-n replied:
Same as `async_set_cooling`

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,108 @@
+    return unload_ok
+
+
+async def update_listener(hass: HomeAssistant, entry: ConfigEntry):
```
Please type the whole signature when adding type annotations.
```suggestion
async def update_listener(hass: HomeAssistant, entry: ConfigEntry) -> None:
```

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,108 @@
+UPDATE_INTERVAL = timedelta(seconds=60)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
```
```suggestion
async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,108 @@
+    ) as err:
+        raise exceptions.ConfigEntryNotReady from err
+
+    base_uh = Alpha2BaseUpdateHandler(hass, base)
```
```suggestion
    coordinator = Alpha2BaseCoordinator(hass, base)
```

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,108 @@
+
+    base_uh = Alpha2BaseUpdateHandler(hass, base)
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {"connection": base_uh, "devices": set()}
```
```suggestion
    hass.data[DOMAIN][entry.entry_id] = coordinator
```

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,108 @@
+    """Set up a config entry."""
+    base = Alpha2Base(entry.data["host"])
+    try:
+        await base.update_data()
```
We can use `coordinator.async_config_entry_first_refresh` instead. It will raise ConfigEntryNotReady by default if the first update fails.

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,108 @@
+
+    async def async_set_target_temperature(self, heatarea_id, target_temperature):
+        """Set the target temperature of the given heatarea."""
+        _LOGGER.info(
```
Max debug level please.

MartinHjelmare reviewed (changes requested):
We should use our data update coordinator helper here.

j-a-n committed:
Refactor using DataUpdateCoordinator

j-a-n commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,108 @@
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+
+    # Trigger updates at regular intervals.
+    async_track_time_interval(hass, base_uh.async_update, UPDATE_INTERVAL)
```
Hello @MartinHjelmare ,
thank you very much for your review.
Please take a look at my changes.

j-a-n commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,183 @@
+            )
+        )
+
+    def _handle_heatarea_data_updated(self, data):
```
I skipped that. I am now using the data from the coordiantor.

j-a-n committed:
Remove _attr_should_poll

j-a-n committed:
Raise HomeAssistantError on communication error

Catch HTTPError instead of broad except and reraise as HomeAssistantError

j-a-n committed:
Change DataUpdateCoordinator name to alpha2_base

j-a-n commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,108 @@
+    async def async_set_cooling(self, enabled: bool):
+        """Enable or disable cooling mode."""
+        await self.base.set_cooling(enabled)
+        for heatarea in self.base.heatareas:
```
No need to fetch data from device at this point, so just calling the listeners to publish the change.

j-a-n commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,108 @@
+            heatarea_mode,
+        )
+        await self.base.update_heatarea(heatarea_id, {"HEATAREA_MODE": heatarea_mode})
+        await self.async_update()
```
Same as `async_set_cooling`

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,181 @@
+        """Initialize Alpha2 ClimateEntity."""
+        super().__init__(coordinator)
+        # Set the right type for mypy
+        self.coordinator: Alpha2BaseCoordinator = self.coordinator
```
Move the type annotation to a class attribute.
```suggestion
```

j-a-n replied:
Thank you, I did not know that.

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,181 @@
+class Alpha2Climate(CoordinatorEntity, ClimateEntity):
+    """Alpha2 ClimateEntity."""
+
+    target_temperature_step = 0.2
```
```suggestion
    coordinator: Alpha2BaseCoordinator
    target_temperature_step = 0.2
```

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,181 @@
+        super().__init__(coordinator)
+        # Set the right type for mypy
+        self.coordinator: Alpha2BaseCoordinator = self.coordinator
+        self.heatarea_id = heatarea_id
```
We prefer to separate words with underscore in variable names.
```suggestion
        self.heat_area_id = heat_area_id
```

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,181 @@
+    coordinator: Alpha2BaseCoordinator = hass.data[DOMAIN][config_entry.entry_id]
+
+    async_add_entities(
+        Alpha2Climate(coordinator, heatarea_id) for heatarea_id in coordinator.data
```
```suggestion
        Alpha2Climate(coordinator, heat_area_id) for heat_area_id in coordinator.data
```

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,181 @@
+            )
+            self.coordinator.data[self.heatarea_id]["T_TARGET"] = target_temperature
+        except aiohttp.web.HTTPException as http_err:
+            raise HomeAssistantError(
```
Consider moving the exception handling to the coordinator.

j-a-n replied:
Done.

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,181 @@
+
+        try:
+            await self.coordinator.async_set_target_temperature(
+                self.coordinator.data[self.heatarea_id]["ID"], target_temperature
```
```suggestion
                self.heat_area_id, target_temperature
```

j-a-n replied:
whoops

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,181 @@
+        else:
+            self.coordinator.data[self.heatarea_id]["HEATAREA_MODE"] = heatarea_mode
+            if heatarea_mode == 1:
+                if self.coordinator.get_cooling():
```
Store the result in a local variable first. Then check heat area mode.

j-a-n replied:
done

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,181 @@
+
+    async def async_set_preset_mode(self, preset_mode: str) -> None:
+        """Set new operation mode."""
+        heatarea_mode = 0
```
```suggestion
        heat_area_mode = 0
```

j-a-n replied:
I renamed all occurrences of heatarea to heat_area where possible.

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,181 @@
+        data["lock_available"] = self.coordinator.data[self.heatarea_id].get(
+            "LOCK_AVAILABLE", False
+        )
+        data["block_hc"] = self.coordinator.data[self.heatarea_id].get(
```
Please explain what these values in the extra state attributes represent.

j-a-n replied:
is that ok?

j-a-n committed:
Refresh coordinator before setting data

j-a-n committed:
Raise ValueError on invalid heat area mode

j-a-n committed:
Rename heatarea to heat_area

j-a-n committed:
Set type annotation in class attribute

j-a-n committed:
Move coordinator to top

j-a-n committed:
Move exception handling to the coordinator

j-a-n committed:
Use heat_area_id directly

j-a-n committed:
Sore get_cooling() result into local var

j-a-n committed:
Add explanation of status attributes

and remove BLOCK_HC

j-a-n commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,105 @@
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    await coordinator.async_config_entry_first_refresh()
```
Makes sense

j-a-n commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,105 @@
+    ) -> None:
+        """Set the mode of the given heatarea."""
+        # HEATAREA_MODE: 0=Auto, 1=Tag, 2=Nacht
+        assert heatarea_mode in (0, 1, 2)
```
Done

frenck commented on .pre-commit-config.yaml:
```diff
@@ -88,13 +88,13 @@ repos:
-        types: [python]
-        require_serial: true
-        files: ^homeassistant/.+\.py$
+      # - id: mypy
+      #  name: mypy
+      #  entry: script/run-in-env.sh mypy
+      #  language: script
+      #  types: [python]
+      #  require_serial: true
+      #  files: ^homeassistant/.+\.py$
```
This change doesn't seem to be related (and not wished for either 😬 )

j-a-n replied:
Oh shit, sorry for that :open_mouth:.
I will try to rebase my branch.

j-a-n replied:
Oh shit, sorry for that :open_mouth:.
I will try to rebase my branch.

j-a-n replied:
Rebase is done.
This is why I commented out the mypy hook:
```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

homeassistant/components/moehlenhoff_alpha2/__init__.py:52: error: "dict" is not subscriptable  [misc]
homeassistant/components/moehlenhoff_alpha2/__init__.py:52: error: "dict" is not subscriptable, use "typing.Dict" instead  [misc]
homeassistant/components/moehlenhoff_alpha2/__init__.py:65: error: "dict" is not subscriptable, use "typing.Dict" instead  [misc]
Found 3 errors in 1 file (checked 1 source file)
```
If I use `typing.Dict` instead, pylint complains about it.

MartinHjelmare replied:
Add at the top of the module:
```py
from __future__ import annotations
```

j-a-n replied:
Thanks!
One mypy error persists:
```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

homeassistant/components/moehlenhoff_alpha2/__init__.py:54: error: "dict" is not subscriptable  [misc]
Found 1 error in 1 file (checked 1 source file)
```

MartinHjelmare replied:
That should work if you're running Python 3.9+.

j-a-n replied:
I am using the dev container.

j-a-n replied:
Never mind, the issue went away somehow :thinking:

j-a-n committed:
Fix pylint warnings

andyboeh commented:
> Are you talking about the attribute _HEATAREA_MODE_? This attribute is set to 3 if I turn off a room controller.

Yes, exactly. I'm not sure what happens with your integration if it encounters this state? As mentioned, I haven't had the time to migrate to this integration, yet.

> I am trying to get this merged since over one year, so I don't want to add any new features in this PR.

Perfectly understandable. Unfortunately, I never found this PR during this time, it's actually older (and cleaner) than my approach.

j-a-n committed:
from __future__ import annotations

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,169 @@
+        await self.coordinator.async_set_target_temperature(
+            self.heat_area_id, target_temperature
+        )
+        self.coordinator.data[self.heat_area_id]["T_TARGET"] = target_temperature
```
Since we're calling the callbacks before we modify the coordinator data here, the entities won't see this update. Maybe move this update to the coordinator data inside the coordinator?

j-a-n replied:
You are right, much better.

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,169 @@
+            self.heat_area_id, heat_area_mode
+        )
+
+        self.coordinator.data[self.heat_area_id]["HEATAREA_MODE"] = heat_area_mode
```
As above.

MartinHjelmare commented on homeassistant/components/moehlenhoff_alpha2/climate.py:
```diff
@@ -0,0 +1,169 @@
+        """Return the extra state attributes.
+
+        is_locked (bool): Indicates whether the operating lock of the room operating panel is active.
+        lock_available (bool): Specifies whether a code is requested when the room operating unit is unlocked.
```
`lock_available` doesn't explain what this attribute mean. If the attribute means whether or not a lock code is required, we should name it thereafter, eg `lock_code_required`.

I think both these attributes can be moved to separate binary sensor entities since they have a meaning regardless of the state of the climate device.

j-a-n replied:
I will remove extra state attributes, because I think they are rarely needed anyway.

MartinHjelmare commented on tests/components/moehlenhoff_alpha2/test_config_flow.py:
```diff
@@ -0,0 +1,91 @@
+    }
+
+
+async def test_duplicate_error(hass: HomeAssistant):
```
```suggestion
async def test_duplicate_error(hass: HomeAssistant) -> None:
```

MartinHjelmare commented on tests/components/moehlenhoff_alpha2/test_config_flow.py:
```diff
@@ -0,0 +1,91 @@
+    with patch(
+        "moehlenhoff_alpha2.Alpha2Base._fetch_static_data", mock_fetch_static_data
+    ):
+        result = await hass.config_entries.flow.async_init(
```
Add a MockConfigEntry instead with the same host as entered in the flow.

https://github.com/home-assistant/core/blob/76bfbbafe1ef3761c75e397c285e8057db926fe4/tests/common.py#L755

j-a-n replied:
I have updated the tests based on the scaffold template

MartinHjelmare commented on tests/components/moehlenhoff_alpha2/test_config_flow.py:
```diff
@@ -0,0 +1,91 @@
+    with patch(
+        "moehlenhoff_alpha2.Alpha2Base._fetch_static_data", mock_fetch_static_data
+    ):
+        result = await hass.config_entries.flow.async_init(
```
We want to patch the integration setup function on all create entry results to avoid setting up the whole integration. See the scaffold test template for example.

j-a-n committed:
Use Platform Enum

j-a-n commented on homeassistant/components/moehlenhoff_alpha2/__init__.py:
```diff
@@ -0,0 +1,123 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = ["climate"]
```
okay

j-a-n committed:
Move data handling to coordinator

j-a-n committed:
Remove property extra_state_attributes

j-a-n committed:
Add missing annotations

j-a-n committed:
Update moehlenhoff-alpha2 to version 1.1.2

j-a-n committed:
Rework tests based on the scaffold template

j-a-n committed:
Set also heat/cool/day/night temp with target temp

MartinHjelmare commented on tests/components/moehlenhoff_alpha2/test_config_flow.py:
```diff
@@ -2,90 +2,119 @@
-        assert result["reason"] == "already_configured"
+        config_entry.add_to_hass(hass)
 
+        await hass.config_entries.async_setup(config_entry.entry_id)
```
We don't need to set up the integration to test this flow.

MartinHjelmare commented on tests/components/moehlenhoff_alpha2/test_config_flow.py:
```diff
@@ -2,90 +2,119 @@
+        "homeassistant.components.moehlenhoff_alpha2.async_setup_entry",
+        return_value=True,
+    ) as mock_setup_entry:
+        config_entry = MockConfigEntry(
```
We can add the mock config entry outside of the patch context manager.

MartinHjelmare commented on tests/components/moehlenhoff_alpha2/test_config_flow.py:
```diff
@@ -2,90 +2,119 @@
+async def test_form_duplicate_error(hass: HomeAssistant) -> None:
+    """Test that errors are shown when duplicates are added."""
+    with patch("moehlenhoff_alpha2.Alpha2Base.update_data", mock_update_data), patch(
+        "homeassistant.components.moehlenhoff_alpha2.async_setup_entry",
```
We don't need to patch the integration setup function on abort results.

MartinHjelmare commented on tests/components/moehlenhoff_alpha2/test_config_flow.py:
```diff
@@ -2,90 +2,119 @@
-        side_effect=asyncio.TimeoutError,
+        "moehlenhoff_alpha2.Alpha2Base.update_data", side_effect=asyncio.TimeoutError
+    ), patch(
+        "homeassistant.components.moehlenhoff_alpha2.async_setup_entry",
```
This patch isn't needed since the integration won't be setup on error result.

MartinHjelmare commented on tests/components/moehlenhoff_alpha2/test_config_flow.py:
```diff
@@ -2,90 +2,119 @@
-        side_effect=Exception,
+        "moehlenhoff_alpha2.Alpha2Base.update_data", side_effect=Exception
+    ), patch(
+        "homeassistant.components.moehlenhoff_alpha2.async_setup_entry",
```
As above.

j-a-n committed:
Remove unneeded code from tests

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
Please link a docs PR in this PR description.

j-a-n commented:
Thanks @MartinHjelmare  for helping me through :pray:

j-a-n commented:
Hi @MartinHjelmare, could you please have a look at the linked documentation PR? Thanks!

MartinHjelmare commented:
You should have pinged me earlier. There's no notification when a PR description is updated. 👍

j-a-n commented:
Thanks a lot @MartinHjelmare 🎉
